### ISSUE TITLE
Compile failure ID from multiple regex groups

### ISSUE THREAD

**Issue by jerrykan** [fail2ban/fail2ban#2755]
I would like to use fail2ban to place temporary blocks on misbehaving clients connect to a authenticating squid proxy. An example of the sort of squid log I'm look at is (with some extra empty lines for clarity):
```
1591843992.923      0 192.168.1.30 TCP_DENIED/407 3988 CONNECT www.example.com:443 - HIER_NONE/- text/html
1591843993.230    432 192.168.1.30 TCP_MISS/200 5438 CONNECT www.example.com:443 - HIER_DIRECT/93.184.216.34 -

1591843994.836      0 192.168.1.50 TCP_DENIED/407 3789 CONNECT www.example.com:443 - HIER_NONE/- text/html
1591843995.833    234 192.168.1.50 TCP_MISS/200 2309 CONNECT www.example.com:443 - HIER_DIRECT/93.184.216.34 -

1591843996.234      0 192.168.1.40 TCP_DENIED/407 3872 CONNECT www.example.com:443 - HIER_NONE/- text/html
1591843997.983    143 192.168.1.40 TCP_MISS/200 4098 CONNECT www.example.com:443 - HIER_DIRECT/93.184.216.34 -

1591843998.143      0 192.168.1.50 TCP_DENIED/407 3872 CONNECT fail.example.com:443 - HIER_NONE/- text/html
1591843998.168      0 192.168.1.50 TCP_DENIED/407 3872 CONNECT fail.example.com:443 - HIER_NONE/- text/html
1591843998.189      0 192.168.1.50 TCP_DENIED/407 3872 CONNECT fail.example.com:443 - HIER_NONE/- text/html
1591843998.202      0 192.168.1.50 TCP_DENIED/407 3872 CONNECT fail.example.com:443 - HIER_NONE/- text/html
1591843999.233      0 192.168.1.50 TCP_DENIED/407 3872 CONNECT fail.example.com:443 - HIER_NONE/- text/html
1591843999.245      0 192.168.1.50 TCP_DENIED/407 3872 CONNECT fail.example.com:443 - HIER_NONE/- text/html
1591843999.261      0 192.168.1.50 TCP_DENIED/407 3872 CONNECT fail.example.com:443 - HIER_NONE/- text/html
1591843999.282      0 192.168.1.50 TCP_DENIED/407 3872 CONNECT fail.example.com:443 - HIER_NONE/- text/html
1591843999.299      0 192.168.1.50 TCP_DENIED/407 3872 CONNECT fail.example.com:443 - HIER_NONE/- text/html
1591843999.309      0 192.168.1.50 TCP_DENIED/407 3872 CONNECT fail.example.com:443 - HIER_NONE/- text/html
```
The first few lines show an example of a normal client (ie. web browser) trying to make a request to the `www.example.com` server:
- client makes a request to the proxy without any authentication credentials
  - proxy responses with a `TCP_DENIED/407`
- client makes the request again, but this time with authentication credentials
  - proxy accepts the authentication credentials, processes the request, and responds with a `TCP_MISS/200`

The last batch of lines represents an example of a misbehaving client trying to connect to `fail.example.com`:
- client makes a request to the proxy without any authentication credentials
  - proxy responses with a `TCP_DENIED/407`
- client ignores the request for authentication credentials and make another request to the proxy without any authentication credentials
  - proxy responses with a `TCP_DENIED/407`
- (this continues over and over until the logs fill up and the runs the server out of diskspace)

---

Looking at this example it is hopefully clear that not all `TCP_DENIED/407` entries are equal. 

What I would like to detect is the instances where there are persistent failed attempts from a client to the same server. To me this would seem to require some way to combine the client IP and the remote server to create a failure id (ie. `192.168.1.50,fail.example.com:443`). For example the above example log would results in the following failure counts:
```
192.168.1.30,www.example.com:443    1
192.168.1.40,www.example.com:443    1
192.168.1.50,www.example.com:443    1
192.168.1.50,fail.example.com:443  10
```
If this sort of failure ID possible I could easily detect that `192.168.1.50,fail.example.com:443` was generating a lot of unexpected failures, and temporary block the client `192.168.1.50`.

**Comment by jerrykan** [comment]
_This issue somewhat a continuation of the discussion from https://github.com/fail2ban/fail2ban/issues/67#issuecomment-642002829 which may provide some extra context to these responses_

---

>So basically it is quasi impossible to obtain multiple IDs from single line (excepting the above mentioned case with DNS).

I'm not necessarily after a way to obtain multiple IDs from a single line, more I'm after a way to combine multiple fields into a single ID.

> Although we have introduced capturing alternate groups with alt_ prefix, but it will neither affect ID of the ticket, nor expand the ticket for every alternate value found.

What might be interesting is if groups could have a `fid_` prefix and the failure ID is created by combining those group values together (either as a tuple or using some sort of string delimiter)

>> I would like to match on both the client IP and the requested domain as a unique pair and ban the IP if there are persistent TCP_DENIED/407 for the same domain.
>
> It looks like you want implement own multi-line handling identified failure.
> So the IP is a ticket ID, and IP with DNS is the multi-line failure ID.
I guess it is hardly possible, but you should provide log excerpt (with several messages) illustrating a failure and normal case (that should be not detected as a failure).
>
> As an example see our current sshd-filter (multiline failure ID <F-MLFID>...</F-MLFID> identifying the session and <HOST> identifying ticket that gets banned). But as already said, it depends. So provide the log-excerpt either.

I couldn't really find any documentation on how `<F-MLFID>` works, nor could I really figure out what it was supposed to be doing in the `sshd-filter`. Based on the name though I suspect it is for filtering based on information in multiple different lines... which wouldn't really seem to help in my use-case, because everything I need is contained on a single line.

> But normally to avoid banning of whole proxy, so to ban proxying requests properly (again not the proxy but rather real-IP (X-Forwarded-For) or session or something identifying user), one have to implement custom action that would prohibit it (e. g. add to blacklist) inside of the proxy (not using net-filter like iptables).

I think there might be some confusion here. I'm not banning any proxies, just the IPs of client's trying to access the proxy. I'm not worried about allowing some requests through and blocking others - if a client is making dodgy requests, I'm happy to block all requests from that IP for a short time. This can be achieved using the existing iptables actions. The problem is coming up with a filter/failure ID to detect when a problem is occurring.

**Comment by sebres** [comment]
Hmm... the idea with `<F-MLFID>` would not work probably, because at the moment it implies that some regex (here matching `TCP_DENIED/407`) would be marked as `<F-NOFAIL>` to signal it is not directly a failure. At the same time it expects that other regex (matching something other as success and also not a first failure) which would additionally signal an ultimate failure case (like "too many" or "connection closed") in order to recognize it was done without success, but it is simply not given. 
So we have a "failure" message with success message hereafter and "failure" message(s) without success message. This scenario is simply impossible at the moment (as solution not as workaround).

One thing that can probably help here would be #1243, but...
But if you read https://github.com/fail2ban/fail2ban/pull/1243#issuecomment-304626176 there are many questions are open.
The feature where single message lets completely forget all previous failures may be very dangerous, for instance assume following scenario:
- an evildoer (that also has a legitimate access) makes several failures (up to `maxretry - 1`);
- then he doing single access with its own (legitimate) credential, in order to force fail2ban to forget failures from above;
- and repeat the cycle again, so he could continue bruteforce attack without a ban, furthermore it'd additionally produce certain (parasitic) load on fail2ban side.

I guess your case requires some special solution (e. g. some pre-processing service counting failures without success).
As long as it is not there I can only suggest an old multi-line filter (using `maxlines` to scan some multi-line window traversing over the log-file):
```ini
[Definition]
maxlines = 2
failregex = ^\s*\d+\s+(?P<__ID><ADDR>)\s+TCP_DENIED/407\s+\d+\s+CONNECT\s+[^\n]*\n(?=\s+\d+ (?P=__ID)\s+TCP_DENIED/407)
```
<details><summary>PoC</summary>

```
$ fail2ban-regex -o row gh-2755.txt gh-2755.conf
['192.168.1.50',        1591843998.168, {'ip6': None, '__ID': u'192.168.1.50', 'ip4': u'192.168.1.50'}],
['192.168.1.50',        1591843998.189, {'ip6': None, '__ID': u'192.168.1.50', 'ip4': u'192.168.1.50'}],
['192.168.1.50',        1591843998.202, {'ip6': None, '__ID': u'192.168.1.50', 'ip4': u'192.168.1.50'}],
['192.168.1.50',        1591843999.233, {'ip6': None, '__ID': u'192.168.1.50', 'ip4': u'192.168.1.50'}],
['192.168.1.50',        1591843999.245, {'ip6': None, '__ID': u'192.168.1.50', 'ip4': u'192.168.1.50'}],
['192.168.1.50',        1591843999.261, {'ip6': None, '__ID': u'192.168.1.50', 'ip4': u'192.168.1.50'}],
['192.168.1.50',        1591843999.282, {'ip6': None, '__ID': u'192.168.1.50', 'ip4': u'192.168.1.50'}],
['192.168.1.50',        1591843999.299, {'ip6': None, '__ID': u'192.168.1.50', 'ip4': u'192.168.1.50'}],
['192.168.1.50',        1591843999.309, {'ip6': None, '__ID': u'192.168.1.50', 'ip4': u'192.168.1.50'}],
```
</details>
But this solution is ugly - let alone it expects at least 2 failures (and ignores the first in such chunk), just try to insert some lines in-between and it would stop to work properly, so imagine you'd get some interim lines logged between `407` and `200` or between `407` and `407` from the same IP (e. g. other concurrent connections, evildoers makes a pause intentionally etc. pp).
So you can miss many failures as well as can get some false positives from legitimate users.

Increasing of the buffer window (`maxlines`) together with `<SKIPLINES>` tag wouldn't help too, because you can not increase it too large (filter will be really slow) and it will be much simpler to catch a false positive (because one must consider every `200` to every `407` in the lines you may skip). I don't see any possibility to implement good filter considering 407/200 in pairs (with some lines in-between) in this way.

Also note that many clients (browsers) could create concurrent connections, so one would see several simultaneous attempts from same IP in unpredictable order.

The logging occurring in this squid proxy is simply not suitable at the moment to implement proper filter.

**Comment by jerrykan** [comment]
> Also note that many clients (browsers) could create concurrent connections, so one would see several simultaneous attempts from same IP in unpredictable order.

I think this gets to the heart of the problem. Adding the fact that there could also potentially be a very large number of client hosts also generating a lot of log information for different IPs, I don't think any sort of multi-line parsing or counter resets (as mentioned in #1243) is viable.

> The logging occurring in this squid proxy is simply not suitable at the moment to implement proper filter.

For our use-case we don't necessarily need fail2ban to have a full understanding of when a failed attempt has resulted in a successful attempt, we just need to know when there are persistent failures of the same type (ie. a specific IP is consistently failing to authenticate for a specific remote server).

Putting aside _how_ it would be implemented, are there any concerns about the functionality suggested in this issue being added to fail2ban (ie. have the ability to combine multiple regex groups together to generate a failure ID for tracking these sorts of failures)?

**Comment by sebres** [comment]
> I don't think any sort of multi-line parsing or counter resets (as mentioned in #1243) is viable.

Well it depends.

> ie. a specific IP is consistently failing to authenticate for a specific remote server

This is basically the "heart of the problem". And "specific remote server" is secondary here, because it is not necessarily safe, for instance if you would see:
```
192.0.2.1 TCP_DENIED/407 3988 CONNECT host1:443
192.0.2.1 TCP_DENIED/407 3988 CONNECT host2:443
...
192.0.2.1 TCP_DENIED/407 3988 CONNECT hostN:443
```
your second ID would be different, but... it doesn't change anything - the evildoer either trying to brute N servers (repeat it later again)... or hostN may be wwwN.example.com (so other host same server or cluster).
So if we'd simply forget this "specific remote server" and remain by "a specific IP is consistently failing to authenticate", it would change nothing.

But, how you want to distinguish this:
```
192.0.2.1 TCP_DENIED/407 3988 CONNECT ...
192.0.2.1 TCP_DENIED/407 3988 CONNECT ...
...
192.0.2.1 TCP_DENIED/407 3988 CONNECT ...
```
from this:
```
192.0.2.1 TCP_DENIED/407 3988 CONNECT ...
192.0.2.1 TCP_DENIED/407 3988 CONNECT ...
...
192.0.2.1 TCP_DENIED/407 3988 CONNECT ...
...
192.0.2.1 TCP_MISS/200 5438 CONNECT ...
192.0.2.1 TCP_MISS/200 5438 CONNECT ...
...
192.0.2.1 TCP_MISS/200 5438 CONNECT ...
```
based on the premise of unpredictable order (they can be mixed), latency (some are retarded up to minutes) and server/client behavior (or conditions like network speed and stability, server load etc).
So for example try to answer the questions:
1. how long can be delay between 407 and 200 to consider the first as a failure (to distinguish from success)
2. is 200 always necessary (or rather under which circumstances can it happen that not every 407 will be followed by 200).
3. why the second ID (host) is interesting at all, if some IP making too many 407

> have the ability to combine multiple regex groups together to generate a failure ID

Sure, it is possible. But for above mentioned reasons I guess it would not help by "tracking these sorts of failures".

**Comment by jerrykan** [comment]
> This is basically the "heart of the problem". And "specific remote server" is secondary here, because it is not necessarily safe, for instance if you would see:
> 
> ```
> 192.0.2.1 TCP_DENIED/407 3988 CONNECT host1:443
> 192.0.2.1 TCP_DENIED/407 3988 CONNECT host2:443
> ...
> 192.0.2.1 TCP_DENIED/407 3988 CONNECT hostN:443
> ```
> 
> your second ID would be different, but... it doesn't change anything - the evildoer either trying to brute N servers (repeat it later again)... or hostN may be wwwN.example.com (so other host same server or cluster).
> So if we'd simply forget this "specific remote server" and remain by "a specific IP is consistently failing to authenticate", it would change nothing.

The `TCP_DENIED/407` response means one of two things. Either there was no authentication credentials supplied, or authentication credentials were supplied but they were not valid.

If we look at the above log entries in the context that all of these requests contained authentication credentials, then we could assume it was an attempt by a user to brute force a username/password. It would not be very successful because the authentication backend should lock the account after a certain number of failed attempts. So this is not a use-case I would need fail2ban to handle.

If we look at the above log entries in the context where none of these requests contain authentication credentials, then it would just seem to be an attempt by a user to find a remote server that the proxy does not require authentication to get to. At best this is just a _very_ inefficient way of finding what access controls the proxy has in place, and at worst a denial-of-service attempt. Neither of these are great, but both assume intentionally bad actors. These sort of bad actors are not the specific use-case I'm interested in.

My use-case is more concerned with unintentionally misbehaving applications/services on client hosts. The answer to question 3 below hopefully explains this...

> So for example try to answer the questions:
> 
>     1. how long can be delay between 407 and 200 to consider the first as a failure (to distinguish from success)
There is no stateful relationship between the two log entries. The only relationship that might be _implied_ between the two requests (a `407` and a `200`) is that a client host may have tried to make a request that wasn't successful (`407`) - either because there were no authentication credentials, or they were invalid - and then at some point later the same client host tried successfully (`200`). However there is no guarantee that those two events are related, or even initiated by the same application on the client host. So trying to match up `407` response with `200` response probably isn't necessarily going to be that helpful. Some assumptions could be made when trying to match them up, but they will just be assumptions... and probably end up be wrong ones.

>     2. is 200 always necessary (or rather under which circumstances can it happen that not every 407 will be followed by 200).

No, a `200` is not always necessary or expected. There may be a background service on a client host that is trying to get out via the proxy, but gets a `407` and doesn't know how to authenticate, so won't try another request. This is the system working as expected and not something I would be concerned about.

>     3. why the second ID (host) is interesting at all, if some IP making too many 407

Often I'm seeing an application or background service on a client host attempting to get out via the proxy to a specific endpoint but getting a `407` in response. It seems the application doesn't get the response it expects and for whatever reason just tries the same request again over and over forever in very quick succession. This is usually not obvious to the user on the client host because it is happening in the background and there are other applications (ie. web browser) on the same client host making both `407` and `200` attempts to get successfully out via the proxy.

If I were to use fail2ban to just block any client host (ie. IP) that has 100 `407` responses within 1 second, then that may be a false positive because there is no indication whether the 100 requests were all to the same endpoint, or different endpoints. If they were all to different endpoints then we can probably make a reasonable assumption that a user has just started up their browser and all the tabs from a previous session are reloading (a bunch of successful `200` responses would likely follow soon after). However if all those 100 requests were to the same endpoint, then it raises some flags that some service/application on a client hosts is misbehaving and needs to be dealt with.

> > have the ability to combine multiple regex groups together to generate a failure ID
> 
> Sure, it is possible. But for above mentioned reasons I guess it would not help by "tracking these sorts of failures".

From my perspective combining regex groups is the only way to detect the sort of use-case I mentioned in the answer to question 3.

**Comment by sebres** [comment]
>  If they were all to different endpoints then we can probably make a reasonable assumption that a user has just started up their browser and all the tabs from a previous session are reloading 

Or the brute-force simply takes place for a lot of end-points in the cycle. 
Don't hold the evil side as dummy - they are smart enough to estimate your `maxretry` (and other params) and would circumvent your jail as they retrieving that.

> However if all those 100 requests were to the same endpoint, then it raises some flags that some service/application on a client hosts is misbehaving and needs to be dealt with.

No... it can also be also a signal for:
- slow (unstable) internet connect to the user (bad mobile net etc) getting fast now;
- end-point had high load for some previous point and now becomes free and answers rapidly;
- browser cache is invalidated (can have also multiple reasons) and several resources going requested at once;
- impossibility to hold on a keep-alive connect on client (unstable, etc) or server side (early washout as expired by too many users);
- a lot of URI's to same endpoint are called (or even few URI's causing loading of resources with new `last-modification`/`etag`);
- dancing on the keyword cat, periodically touching F5;
- etc pp;

I stay by my opinion - implementing of such filter is hardly possible (regardless the second ID is used or not), at least as long as logging format doesn't change (would not provide more information or more lines).

**Comment by jerrykan** [comment]
> > If they were all to different endpoints then we can probably make a reasonable assumption that a user has just started up their browser and all the tabs from a previous session are reloading
> 
> Or the brute-force simply takes place for a lot of end-points in the cycle.
> Don't hold the evil side as dummy - they are smart enough to estimate your `maxretry` (and other params) and would circumvent your jail as they retrieving that.

There seems to be an assumption that there is an intentionally bad actor at play, which is a valid use-case, but it is not the specific use-case I am interested in in this instance.

> > However if all those 100 requests were to the same endpoint, then it raises some flags that some service/application on a client hosts is misbehaving and needs to be dealt with.
> 
> No... it can also be also a signal for:
>  ...

Yes it _could_ be a signal for lots of things, but in our environment most of these possibilities are not necessarily applicable. We can't necessarily predict every eventuality, but fail2ban is a very flexible and useful tool to help identify potential problems and solve a real issue we are seeing.

> I stay by my opinion - implementing of such filter is hardly possible (regardless the second ID is used or not), at least as long as logging format doesn't change (would not provide more information or more lines).

I'm confused. What part of "implementing such filter" is hardly possible? The ability to implement in fail2ban the functionality of combining regex group values into a failure ID?

**Comment by sebres** [comment]
> in our environment most of these possibilities are not necessarily applicable

Ok. I though we ware speaking about normal case (proxying in the wild internet).

> The ability to implement in fail2ban the functionality of combining regex group values into a failure ID?

Nope. Implementing of the filter to consider 407 as failure (in some magic cases) and don't (in other  magic cases). I understand in about how you would handle that, but still doubt it'd be correct.
The ability to conjoin some IDs to single one is of course given. As I wrote above - "Sure, it is possible. But for above mentioned reasons I guess it would not help ...".

**Comment by sebres** [comment]
Basically, I have at all no time at the moment for such RFE, but...
I had an idea how one could extend `alt_`-prefix mimic to implement that, so trying to summarize that, I implemented it in dd8081ade5ace23782efcf0c96b7f8c9f16228a7 (as `tuple_` prefix in filter capture groups).

Some thoughts:
- I have no interest to catch compat issues in 0.10th version, so if we'll provide it, it'd be a part of 0.11, which is released also, but not so widespread as 0.10 (but one can cherry-pick it or create a patch for 0.10 if one'd need);
- I have still not tested it in server, together with action and DB (tested only with `fail2ban-regex`)... 
therefore two points additionally: 
  * I cannot exclude that some explicit string conversion may be expected somewhere (but hope it would work as is);
  * I didn't analyzed how it would affect family recognition (e. g. usage of `<F-ID>` as tuple together with `<ADDR>` and `<HOST>` tags could bother to distinguish IPv6 and IPv4); it can obtain the family from IP, but I cannot exclude that something can still use ticket ID somewhere.

Here are examples explaining that:
```bash
$ fail2ban-regex -o id  '1591983743.667 192.0.2.1 test' '^\s*<F-ID/> <F-TUPLE_ID>\S+</F-TUPLE_ID>'
('192.0.2.1', 'test')
$ fail2ban-regex -o row '1591983743.667 192.0.2.1 test' '^\s*<F-ID/> <F-TUPLE_ID>\S+</F-TUPLE_ID>'
[('192.0.2.1', 'test'), 1591983743.667, {'tuple_id': 'test', 'fid': ('192.0.2.1', 'test')}],

$ fail2ban-regex -o id  '1591983743.667 left [192.0.2.4]:12345 right' '^\s*<F-TUPLE_ID_1>\S+</F-TUPLE_ID_1> <F-ID><ADDR>:<F-PORT/></F-ID> <F-TUPLE_ID_2>\S+</F-TUPLE_ID_2>'
('[192.0.2.4]:12345', 'left', 'right')
$ fail2ban-regex -o row '1591983743.667 left [192.0.2.4]:12345 right' '^\s*<F-TUPLE_ID_1>\S+</F-TUPLE_ID_1> <F-ID><ADDR>:<F-PORT/></F-ID> <F-TUPLE_ID_2>\S+</F-TUPLE_ID_2>'
[('[192.0.2.4]:12345', 'left', 'right'),        1591983743.667, {'fid': ('[192.0.2.4]:12345', 'left', 'right'), 'fport': '12345', 'tuple_id_1': 'left', 'tuple_id_2': 'right', 'ip6': None, 'ip4': '192.0.2.4'}],
```
And I've still no idea how it would help you to implement good filter...
Either I can imagine that it'd explode (somewhere in fail-manager) with a couple of tickets instead of single tickets previously.
But I've already expressed my objections to that.

**Comment by jerrykan** [comment]
This looks interesting. I'll take a look.

> Either I can imagine that it'd explode (somewhere in fail-manager) with a couple of tickets instead of single tickets previously.

Hmmm... this is something I hadn't considered, I assume there would be memory implications associated with this. Would having a shorter `findtime` help to reduce this impact?

**Comment by sebres** [comment]
> Would having a shorter findtime help to reduce this impact?

Sure... the question is who is "faster" - clients flooding it (regardless malignancy of them) or fail2ban removing tickets exceeding the `findtime`.
Anyway I would also reduce `maxmatches` (up to 0), see #2402.

**Comment by sebres** [comment]
merged in 9d076af9a26a81e79d4b02b57eef634bf90e2bd1
